File: src/test/java/org/radarcns/hdfs/data/CompressionFactoryTest.java
package org.radarcns.hdfs.data;

import static java.nio.charset.StandardCharsets.US_ASCII;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class CompressionFactoryTest {

    private CompressionFactory factory;

    @BeforeEach
    public void setUp() {
        factory = new CompressionFactory();
    }

    @Test
    public void testIdentity() throws IOException {
        assertSame(IdentityCompression.class, factory.get("none").getClass());
        assertSame(IdentityCompression.class, factory.get("identity").getClass());

        Compression compression = factory.get("none");

        assertEquals("", compression.getExtension());

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        OutputStream compressedStream = compression.compress("something", out);
        compressedStream.write("abcdef".getBytes(US_ASCII));
        compressedStream.close();

        assertEquals("abcdef", new String(out.toByteArray(), US_ASCII));

        ByteArrayInputStream in = new ByteArrayInputStream("abcdef".getBytes(US_ASCII));
        InputStream compressedIn = compression.decompress(in);
        byte[] result = new byte[10];
        assertEquals(6, compressedIn.read(result));
        assertEquals("abcdef", new String(result, 0, 6, US_ASCII));
    }

    @Test
    public void testGzip() throws IOException {
        assertSame(GzipCompression.class, factory.get("gzip").getClass());

        Compression compression = factory.get("gzip");

        assertEquals(".gz", compression.getExtension());

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        OutputStream compressedStream = compression.compress("something", out);
        compressedStream.write("abcdef".getBytes(US_ASCII));
        compressedStream.close();

        ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
        InputStream compressedIn = compression.decompress(in);
        byte[] result = new byte[10];
        assertEquals(6, compressedIn.read(result));
        assertEquals("abcdef", new String(result, 0, 6, US_ASCII));
    }


    @Test
    public void testZip() throws IOException {
        assertSame(ZipCompression.class, factory.get("zip").getClass());

        Compression compression = factory.get("zip");

        assertEquals(".zip", compression.getExtension());

        ByteArrayOutputStream out = new ByteArrayOutputStream();
        OutputStream compressedStream = compression.compress("something", out);
        compressedStream.write("abcdef".getBytes(US_ASCII));
        compressedStream.close();

        ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
        InputStream compressedIn = compression.decompress(in);
        byte[] result = new byte[10];
        assertEquals(6, compressedIn.read(result));
        assertEquals("abcdef", new String(result, 0, 6, US_ASCII));
    }
}

File: src/test/java/org/radarbase/output/data/CompressionFactoryTest.kt
package org.radarbase.output.data

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.radarbase.output.compression.Compression
import org.radarbase.output.compression.CompressionFactory
import org.radarbase.output.compression.GzipCompression
import org.radarbase.output.compression.IdentityCompression
import org.radarbase.output.compression.ZipCompression
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.charset.StandardCharsets.US_ASCII

class CompressionFactoryTest {

    private lateinit var factory: CompressionFactory

    @BeforeEach
    fun setUp() {
        factory = CompressionFactory()
    }

    @Test
    @Throws(IOException::class)
    fun testIdentity() {
        assertSame(IdentityCompression::class.java, factory["none"].javaClass)
        assertSame(IdentityCompression::class.java, factory["identity"].javaClass)

        val compression = factory["none"]

        assertEquals("", compression.extension)

        val out = ByteArrayOutputStream()
        val compressedStream = compression.compress("something", out)
        compressedStream.write("abcdef".toByteArray(US_ASCII))
        compressedStream.close()

        assertEquals("abcdef", String(out.toByteArray(), US_ASCII))

        val `in` = ByteArrayInputStream("abcdef".toByteArray(US_ASCII))
        val compressedIn = compression.decompress(`in`)
        val result = ByteArray(10)
        assertEquals(6, compressedIn.read(result))
        assertEquals("abcdef", String(result, 0, 6, US_ASCII))
    }

    @Test
    @Throws(IOException::class)
    fun testGzip() {
        assertSame(GzipCompression::class.java, factory["gzip"].javaClass)

        val compression = factory["gzip"]

        assertEquals(".gz", compression.extension)
        testReCompression(compression)
    }

    private fun testReCompression(compression: Compression) {
        val compressed = ByteArrayOutputStream().use { out ->
            compression.compress("something", out).use {
                it.write("abcdef".toByteArray(US_ASCII))
            }
            out.toByteArray()
        }

        val result = ByteArray(10)
        val numBytes = ByteArrayInputStream(compressed).use { `in` ->
            compression.decompress(`in`).use {
                it.read(result)
            }
        }
        assertEquals(6, numBytes)
        assertEquals("abcdef", String(result, 0, 6, US_ASCII))
    }

    @Test
    @Throws(IOException::class)
    fun testZip() {
        assertSame(ZipCompression::class.java, factory["zip"].javaClass)

        val compression = factory["zip"]

        assertEquals(".zip", compression.extension)

        testReCompression(compression)
    }
}
